blockprod/src/detail/mod.rs

```diff
@@ -180,19 +180,51 @@ impl BlockProduction {
         &self,
         current_tip: Id<GenBlock>,
         min_block_timestamp: BlockTimestamp,
+        transactions: Vec<SignedTransaction>,
+        transaction_ids: Vec<Id<Transaction>>,
+        include_mempool: bool,
     ) -> Result<Option<Box<dyn TransactionAccumulator>>, BlockProductionError> {
-        let max_block_size = self.chain_config.max_block_size_from_std_scripts();
+        let mut accumulator: Box<dyn TransactionAccumulator + Send> =
+            Box::new(DefaultTxAccumulator::new(
+                self.chain_config.max_block_size_from_std_scripts(),
+                current_tip,
+                min_block_timestamp,
+            ));
+
+        for transaction in transactions.into_iter() {
+            let transaction_id = transaction.transaction().get_id();
+            // TODO: fees
+            accumulator
+                .add_tx(transaction, Amount::ZERO.into())
```

Comment:
@iljakuklic I wasn't too sure what the fees should be here?

Reply:
> Since the transaction sequence is already passed in, I don't see why the mempool stuff should be involved at all.

By having the option of supplied transactions, supplied mempool transaction Ids, and also fill from random mempool transactions, I think that should cover all scenarios that a user may want... but yes, this does make dependency resolution a bit messy.

Doing it outside of mempool is doable, but mempool's `collect_txs()` function does a few handy things that I would need to duplicate i.e. timelocks, parent/child resolution etc.

What are your thoughts about the way to go about the design? I was thinking I could split `collect_txs()` into three functions - one internal that does all the work, one to keep the same interface as we have now for the current callers, and another which blockprod uses (which passes in a `Vec<Transaction>`, creates a tx verifier, then calls the internal function), and so there is less duplication... unless I'm misunderstanding and there is a better way to go about this?

Reply:
Thinking about this from another angle, for the first two parameters, the caller is already signalling their desired transaction sequence (both supplied and mempool txid), and as this is explicit, we should expect they have already done the resolution. So I just need to use a tx verifier here and gather within an accumulator. Then I can pass this into `collect_txs()` which will then fill the rest of the space with random mempool transactions.

Does this makes better sense to you?

Reply:
> Be aware that the accumulator already assumes the transactions have been verified by the mempool. Fees are extracted from the verifier.

Looking at the transaction verifier, I'm not sure that this is going to be a problem unless I'm mistaken - the is only used to build the block which is then returned to the caller. They still need to submit the block, which will then have to go through the transaction verifier on the round trip. Thoughts?

Reply:
> > Since the transaction sequence is already passed in, I don't see why the mempool stuff should be involved at all.
> 
> By having the option of supplied transactions, supplied mempool transaction Ids, and also fill from random mempool transactions, I think that should cover all scenarios that a user may want... but yes, this does make dependency resolution a bit messy.
> 
> Doing it outside of mempool is doable, but mempool's `collect_txs()` function does a few handy things that I would need to duplicate i.e. timelocks, parent/child resolution etc.
> 
> What are your thoughts about the way to go about the design? I was thinking I could split `collect_txs()` into three functions - one internal that does all the work, one to keep the same interface as we have now for the current callers, and another which blockprod uses (which passes in a `Vec<Transaction>`, creates a tx verifier, then calls the internal function), and so there is less duplication... unless I'm misunderstanding and there is a better way to go about this?

Ok, I see your points here. Probably the most straightforward, albeit probably not the cleanest, would be to pass an additional argument to `collect_txs` specifying all the "extra" transactions the user wants. It would then use these transactions as a matter of priority. We can see how bad it is and refine it further. Maybe the transaction accumulator machinery could be split off and generalized to support multiple "sources", mempool being one, user-supplied transactions being another.

Also note that mempool already has some information regarding transactions given by ID, i.e. the dependency info and fee rates. When looking up transaction by ID, you only get a `SignedTransaction`, so that information is lost. It is, however, quite useful while constructing blocks.

Reply:
So after some digging in order to find how the fees are set, I've found that the verifier needs the store (either chain or mempool) - both of which aren't the source for transactions that are provided over the wire...

So I'm now thinking that if we take transactions in hex, we should probably add them to the mempool. That also gives the benefit of only needing to add a "priority_transaction_ids" `Vec` to `collect_txs()`.